### PR TITLE
Add strip API to  Foreign Array

### PR DESCRIPTION
**Pull request by rnjtranjan**
Fix for #1254

### CODE REVIEW

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 1]
```diff
@@ -1597,3 +1599,37 @@ cmp arr1 arr2 =
...
+strip eq arr@Array{..} =
+    let p = unsafeForeignPtrToPtr aStart
+        q = aEnd
+        len = length arr
+        st = getStart p len
+        end = do
+            st' <- st
+            if st' == q
+            then return q
+            else getLast $ q `plusPtr` negate (sizeOf (undefined :: a))
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1614]: Use the macros from ArrayMacros.h to manipulate the pointers.

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 2]
```diff
@@ -1597,3 +1599,37 @@ cmp arr1 arr2 =
...
+    in go st end
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1615]: `go` is not a recursive function. We can inline the definition right here, that would be more readable.

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 3]
```diff
@@ -1597,3 +1599,37 @@ cmp arr1 arr2 =
 instance NFData (Array a) where
     {-# INLINE rnf #-}
     rnf Array {} = ()
+
+strip :: forall a m. (Storable a, MonadIO m) =>
+    (a -> Bool) -> Array a -> m (Array a)
+strip eq arr@Array{..} =
+    let p = unsafeForeignPtrToPtr aStart
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1606]: We will need a `touch` to ensure the pointer is not garbage collected.

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 4]
```diff
@@ -1597,3 +1599,37 @@ cmp arr1 arr2 =
...
+        q = aEnd
+        len = length arr
+        st = getStart p len
+        end = do
+            st' <- st
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1611]: The action `st` is being called at two places, we are effectively searching for start twice, instead we can call it once and use the result.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]
This is based on very old commit, the array type has changed since. Can you rebase and fix?

Can you write a basic test for it?

**Comment by harendra-kumar** [comment]
Can you add a benchmark as well?

**Review by rnjtranjan — COMMENTED** [review]
Looks good.

**Pull request merged**